### PR TITLE
Make exercise "in grading" status change thread safe

### PR DESCRIPTION
**Pull request by nikke234**
This is a minor change to Stella's PR, which includes thread safety, avoiding repeated API requests, and supporting multiple exercises in grading.

### CODE REVIEW

**Comment by nikke234** [comment]
Note, that sonar for some reason has analyzed the whole code base instead of the changes in this PR, so the above warnings are all unrelated to this change

File: src/main/java/fi/aalto/cs/apluscourses/utils/observable/ObservableProperty.java [hunk 1]
```diff
@@ -80,6 +80,14 @@ public T getAndSet(T value) {
     throw new UnsupportedOperationException();
   }
 
+  /**
+   * Notifies all observers with the current value. Works similarly to {@link
+   * ObservableProperty#set}, but the actual value isn't changed.
+   */
+  public void valueChanged() {
+    onValueChanged(get(), null);
+  }
```
Comment (nikke234) [inline src/main/java/fi/aalto/cs/apluscourses/utils/observable/ObservableProperty.java:89]: @OlliKiljunen was way ahead of me here :smile: 
https://github.com/Aalto-LeTech/intellij-plugin/pull/211/files/fcd4f1fb452a58e3f81ec4865a7fefa32207c4b6#r437382799

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java [hunk 2]
```diff
@@ -134,16 +133,36 @@ public Options getExerciseFilterOptions() {
     return exerciseFilterOptions;
   }
 
-  private void setInGrading(List<ExerciseGroup> exerciseGroups) {
-    final long id = submittedForGrading.getId();
+  private static void setInGrading(List<ExerciseGroup> exerciseGroups, long id) {
     exerciseGroups
         .stream()
         .map(ExerciseGroup::getExercises)
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:139]: either `.filter(Objects::nonNull)` here

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java [hunk 3]
```diff
@@ -134,16 +133,36 @@ public Options getExerciseFilterOptions() {
...
+   * Modifies the 'inGrading' status of the exercise in 'exercisesViewModel' which corresponds to
+   * the given exercise (i.e. has the same ID). The given exercise is not modified. The observers of
+   * 'exercisesViewModel' are notified.
+   */
+  public void setSubmittedForGrading(@NotNull Exercise submittedForGrading) {
+    Exercise previous = inGrading.putIfAbsent(submittedForGrading.getId(), submittedForGrading);
+    if (previous == null) {
+      // This method gets called repeatedly by SubmissionStatusUpdater, so we check whether the
+      // entry was already there or not, and only update the tree if it wasn't there previously.
+      setInGrading(exercisesViewModel.get().getModel(), submittedForGrading.getId());
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:154]: perhaps ensure _`nonNull`_ model is returned?

Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:154]: It's luckily annotated with `@NotNull`

**Pull request merged**